### PR TITLE
Update arrow rustdocs (#4071)

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Part of #4071

# Rationale for this change
 

The current docs are out of date following #4061, and the output for some is pretty hard to parse

image

This updates the docs so that the type summaries are a single line and sufficiently descriptive, and adds some new content about array buffers.

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-array/src/lib.rs [hunk 1]
```diff
@@ -19,50 +19,14 @@
 //! all having the same type. This crate provides concrete implementations of each type, as
 //! well as an [`Array`] trait that can be used for type-erasure.
 //!
-//! # Downcasting an Array
```
Comment (tustvold) [inline arrow-array/src/lib.rs:22]: This content is moved later

File: arrow-array/src/lib.rs [hunk 2]
```diff
@@ -91,63 +55,114 @@
...
-//!     array.len(),
-//!     "The array has 5 values, counting the null value"
-//! );
+//! assert_eq!(5, array.len());
+//! assert_eq!(2, array.value(2));
+//! assert_eq!(&array.values()[3..5], &[3, 4])
+//! ```
 //!
-//! assert_eq!(2, array.value(2), "Get the value with index 2");
+//! # Low-level API
```
Comment (tustvold) [inline arrow-array/src/lib.rs:71]: This reworks the internal representation section, taking some inspiration from https://jorgecarleitao.github.io/arrow2/main/guide/low_level.html

File: arrow/src/lib.rs [hunk 3]
```diff
@@ -317,19 +317,6 @@
 //! assert_eq!(string.value(1), "foo");
 //! ```
 //!
-//! # Memory and Buffers
```
Comment (tustvold) [inline arrow/src/lib.rs:320]: This section is outdated and documenting it at this level just adds confusion, lets leave that for the arrow_array docs

File: arrow-array/src/array/struct_array.rs [hunk 4]
```diff
@@ -22,9 +22,9 @@ use arrow_schema::{ArrowError, DataType, Field, FieldRef, Fields, SchemaBuilder}
 use std::sync::Arc;
 use std::{any::Any, ops::Index};
 
-/// A nested array type where each child (called *field*) is represented by a separate
-/// array.
+/// An array of [tuples](https://arrow.apache.org/docs/format/Columnar.html#struct-layout)
```
Comment (alamb) [inline arrow-array/src/array/struct_array.rs:25]: It is an array of structs, right?

```suggestion
/// An array of [structs](https://arrow.apache.org/docs/format/Columnar.html#struct-layout)
```

File: arrow-buffer/src/buffer/scalar.rs [hunk 5]
```diff
@@ -22,12 +22,24 @@ use std::fmt::Formatter;
 use std::marker::PhantomData;
 use std::ops::Deref;
 
-/// Provides a safe API for interpreting a [`Buffer`] as a slice of [`ArrowNativeType`]
+/// A strongly-typed [`Buffer`] supporting zero-copy cloning and slicing
 ///
-/// # Safety
+/// The easiest way to think about `ScalarBuffer<T>` is being equivalent to a `Arc<Vec<T>>`,
```
Comment (alamb) [inline arrow-buffer/src/buffer/scalar.rs:27]: This is a very helpful comparison. Thank you

File: arrow-array/src/lib.rs [hunk 6]
```diff
@@ -91,63 +55,114 @@
...
+//!
+//! Internally, arrays consist of one or more shared memory regions backed by [`Buffer`],
```
Comment (alamb) [inline arrow-array/src/lib.rs:73]: Backed by "a" buffer?

File: arrow-array/src/lib.rs [hunk 7]
```diff
@@ -91,63 +55,114 @@
...
+//!
+//! For example, the type [`Int16Array`] represents an array of 16-bit integers and consists of:
+//!
+//! * An optional [`NullBuffer`] identifying any null values
+//! * A contiguous [`ScalarBuffer<i16>`]
+//!
+//! Similarly, the type [`StringArray`] represents an array of UTF-8 strings and consists of:
+//!
+//! * An optional [`NullBuffer`] identifying any null values
+//! * An offsets [`ScalarBuffer<i32>`] identifying valid UTF-8 sequences within the values buffer
```
Comment (alamb) [inline arrow-array/src/lib.rs:85]: Aren't the offsets actually an `OffsetBuffer`?

File: arrow-array/src/lib.rs [hunk 8]
```diff
@@ -91,63 +55,114 @@
...
+//! This requires downcasting to the concrete type of the array:
 //!
-//! Similarly, the type [`StringArray`] represents an array of UTF-8 strings and consists of:
+//! ```
+//! # use arrow_array::{Array, Float32Array, Int32Array};
 //!
-//! * An optional [`NullBuffer`] identifying any null values
-//! * An offsets [`Buffer`] of 32-bit integers identifying valid UTF-8 sequences within the values buffer
-//! * A values [`Buffer`] of UTF-8 encoded string data
+//! fn sum_int32(array: &dyn Array) -> i32 {
```
Comment (alamb) [inline arrow-array/src/lib.rs:139]: it might help to add some additional context to what the function is doing:

```suggestion
//! // Safely downcast an `Array` to an `Int32Array` and compute the sum
//! // using native i32 values
//! fn sum_int32(array: &dyn Array) -> i32 {
```

File: arrow-array/src/lib.rs [hunk 9]
```diff
@@ -91,63 +55,114 @@
...
+//!     let integers: &Int32Array = array.as_any().downcast_ref().unwrap();
+//!     integers.iter().map(|val| val.unwrap_or_default()).sum()
+//! }
+//!
+//! // Note: the values for positions corresponding to nulls will be arbitrary
```
Comment (alamb) [inline arrow-array/src/lib.rs:144]:
```suggestion
//! // Safely downcasts the array to a `Float32Array` and returns a &[f32] view of the data
//! // Note: the values for positions corresponding to nulls will be arbitrary (but still valid f32)
```

File: arrow-array/src/lib.rs
Comment (alamb) [inline arrow-array/src/lib.rs:1]: I reviewed the content in this module in its rendered form: 
Screenshot 2023-05-10 at 6 29 56 PM

Love it

**Review by alamb — COMMENTED** [review]
Thank you @tustvold  -- I went over these changes carefully and I think they are a great step forward

**Review by alamb — APPROVED** [review]
Thank you @tustvold  -- I went over these changes carefully and I think they are a great step forward

File: arrow-buffer/src/buffer/scalar.rs (hunk 5 above)
Comment (tustvold) [reply arrow-buffer/src/buffer/scalar.rs:27]: I may have copied it from somewhere :innocent:
